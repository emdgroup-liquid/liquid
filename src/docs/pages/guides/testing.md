---
eleventyNavigation:
  key: Testing
  parent: Guides
  order: 10
layout: layout.njk
title: Testing
permalink: guides/testing/
---

# Testing

If you want to __unit test__ code in your app, which depends on the functionality of Liquid Oxygen Web Components, without mocking their internals, then you must make sure all relevant components have _hydrated_ in your test setup, before you execute your test code. You can do so by running the `defineCustomElements()` method imported from `'@emdgroup-liquid/liquid/dist/loader'` before each test and a `await`ing a helper function, which makes sure all components involved in the test have hydrated.

The following example shows how to execute unit tests with hydrated Web Components with [vitest](https://vitest.dev/), React, and the [React Testing Library](https://testing-library.com/docs/react-testing-library/intro/).

First, [configure vitest to use a setup file](https://vitest.dev/config/#setupfiles):

```js
// vite.config.ts
import { defineConfig } from 'vitest/config'
export default defineConfig({
  // ...
  test: {
    // ...
    setupFiles: './src/setupTests.ts',
  },
})
```

The `setupTests.ts` file looks like this:

```js
// src/setupTests.ts
import { defineCustomElements } from '@emdgroup-liquid/liquid/dist/loader'
import '@testing-library/jest-dom'
import matchers from '@testing-library/jest-dom/matchers'
import { cleanup } from '@testing-library/react'
import { vi } from 'vitest'

// Extend Vitest's expect method with methods from react-testing-library.
expect.extend(matchers)

beforeAll(() => {
  // eslint-disable-next-line @typescript-eslint/ban-ts-comment
  // @ts-ignore
  global.__LD_ASSET_PATH__ = '/' // removes CDN warning
  defineCustomElements()

  global.fetch = vi
    .fn()
    .mockReturnValue(Promise.resolve({ text: () => Promise.resolve('') }))
})

// Run a cleanup after each test case.
afterEach(() => {
  cleanup()
})
```

In your test file, you can now wait for the Web Components to hydrate before executing your actual test code.

```tsx
// src/App.test.tsx
import App from './App'
import { render, screen, waitFor, fireEvent } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import React from 'react'

async function waitForLiquidToHydrate() {
  return waitFor(() => {
    Array.from(document.querySelectorAll('*'))
      .filter((e) => e.tagName.startsWith('LD-'))
      .forEach((component) => {
        expect(component).toHaveClass('hydrated')
      })
  })
}

test('validates input', async () => {
  render(<App />)

  await waitForLiquidToHydrate()

  const ldInputName = screen.getByPlaceholderText('e.g. Jason Parse')
  expect(ldInputName).toBeInTheDocument()

  const ldButtonSubmit = screen.getByText('Submit')
  expect(ldButtonSubmit).toBeInTheDocument()
  userEvent.click(ldButtonSubmit)

  const form = await screen.findByTestId('form')
  expect(form).toBeInTheDocument()

  fireEvent.submit(form)

  await screen.findByText('Your full name is required.')
})
```

Check out our [React sandbox app](https://stackblitz.com/github/emdgroup-liquid/liquid-sandbox-react-tailwind?file=src%2FApp.test.tsx) for a full working example.

<docs-page-nav prev-href="guides/design-tokens/" next-title="Sandbox applications" next-href="guides/sandbox-applications/"></docs-page-nav>

