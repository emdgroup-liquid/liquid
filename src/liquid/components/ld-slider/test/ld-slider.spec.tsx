import { h } from '@stencil/core'
import { newSpecPage } from '@stencil/core/testing'
import { LdSlider } from '../ld-slider'

describe('ld-slider', () => {
  describe('renders', () => {
    it('default', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('custom edges', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider max={50} min={-50} stops="-25,0,25" />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('2 thumbs', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider value="10,50" />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('2 thumbs with custom thumb labels', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => (
          <ld-slider value="10,50" labelFrom="Von" labelTo="Bis" />
        ),
      })
      expect(page.root).toMatchSnapshot()
    })

    it('more than 2 thumbs with custom thumb labels', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider value="10,50,80" labelValue="Thumb" />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('with values always visible', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider alwaysShowValues />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('disabled', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider disabled />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('aria-disabled', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider ariaDisabled="true" />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('with inverted color indicators', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider negative />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('with custom radix', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider radix={16} />,
      })
      expect(page.root).toMatchSnapshot()
    })

    it('with custom width', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider width="500px" />,
      })
      expect(page.root).toMatchSnapshot()
    })
  })

  it('re-renders on prop changes', async () => {
    const page = await newSpecPage({
      components: [LdSlider],
      template: () => <ld-slider radix={16} />,
    })

    const slider = page.root as HTMLLdSliderElement
    slider.alwaysShowValues = true
    slider.ariaDisabled = 'true'
    slider.disabled = true
    slider.labelFrom = 'Von'
    slider.labelTo = 'Bis'
    slider.labelValue = 'Thumb'
    slider.max = 99
    slider.min = 1
    slider.negative = true
    slider.radix = 10
    slider.stops = '30,50,70'
    slider.value = '15,85'
    slider.width = '500px'
    await page.waitForChanges()

    expect(page.root).toMatchSnapshot()
  })

  it('updates css props on value changes', async () => {
    const page = await newSpecPage({
      components: [LdSlider],
      template: () => <ld-slider />,
    })
    const input = page.root.shadowRoot.querySelector('input')

    input.value = '50'
    input.dispatchEvent(new Event('input'))
    await page.waitForChanges()

    expect(page.root).toMatchSnapshot()
  })

  it('emits ldchange event on value changes', async () => {
    const page = await newSpecPage({
      components: [LdSlider],
      template: () => <ld-slider />,
    })
    const input = page.root.shadowRoot.querySelector('input')
    const changeHandler = jest.fn()

    page.root.addEventListener('ldchange', changeHandler)
    input.value = '50'
    input.dispatchEvent(new Event('input'))

    expect(changeHandler).toHaveBeenCalledWith(
      expect.objectContaining({ detail: [50] })
    )
  })

  it('prevents invalid values', async () => {
    const page = await newSpecPage({
      components: [LdSlider],
      template: () => <ld-slider value="-1,101" />,
    })

    expect(page.root).toMatchSnapshot()
  })

  describe('strict mode', () => {
    it('prevents thumbs from being swapped', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider strict value="40,60" />,
      })
      const firstInput = page.root.shadowRoot.querySelector('input')

      firstInput.value = '70'
      firstInput.dispatchEvent(new Event('input'))
      await page.waitForChanges()

      expect(page.root).toMatchSnapshot()
    })

    it('corrects invalid initial values', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider strict value="100,50,90,40" />,
      })

      expect(page.root).toMatchSnapshot()
    })

    it('prevents invalid value changes without re-rendering', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider strict value="0,100" />,
      })
      const slider = page.root as HTMLLdSliderElement

      slider.value = '51,50'
      await page.waitForChanges()

      expect(page.root).toMatchSnapshot()
    })

    it('emits ldchange event after correcting invalid value changes', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider strict value="40,60" />,
      })
      const firstInput = page.root.shadowRoot.querySelector('input')
      const changeHandler = jest.fn()

      page.root.addEventListener('ldchange', changeHandler)
      firstInput.value = '70'
      firstInput.dispatchEvent(new Event('input'))

      expect(changeHandler).toHaveBeenCalledTimes(1)
      expect(changeHandler).toHaveBeenCalledWith(
        expect.objectContaining({ detail: [60, 60] })
      )
    })

    it('does not emit ldchange event on invalid value changes', async () => {
      const page = await newSpecPage({
        components: [LdSlider],
        template: () => <ld-slider strict value="0,100" />,
      })
      const slider = page.root as HTMLLdSliderElement
      const changeHandler = jest.fn()

      page.root.addEventListener('ldchange', changeHandler)
      slider.value = '51,50'

      expect(changeHandler).not.toHaveBeenCalled()
    })
  })
})
