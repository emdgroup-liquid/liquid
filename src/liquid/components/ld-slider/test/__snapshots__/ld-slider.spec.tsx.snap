// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ld-slider prevents invalid values 1`] = `
<ld-slider class="ld-slider" role="group" value="0,100" style="--v0: 0; --v1: 100; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v1) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      From
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="0"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <label class="sr-only" htmlfor="v1">
      To
    </label>
    <input class="ld-slider__input" id="v1" max="100" min="0" type="range" value="100"><output class="ld-slider__output" htmlfor="v1" style="--c: var(--v1);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider re-renders on prop changes 1`] = `
<ld-slider class="ld-slider ld-slider--disabled" role="group" value="15,85" style="--v0: 15; --ld-slider-width: 500px; --min: 1; --max: 99; --fill: 
linear-gradient(
  90deg,
  red 100%,
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v1) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
); --v1: 85;">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Von
    </label>
    <input aria-disabled="true" class="ld-slider__input" disabled="" id="v0" max="99" min="1" type="range" value="15"><output class="ld-slider__output ld-slider__output--permanent" htmlfor="v0" style="--c: var(--v0);"></output>
    <label class="sr-only" htmlfor="v1">
      Bis
    </label>
    <input aria-disabled="true" class="ld-slider__input" disabled="" id="v1" max="99" min="1" type="range" value="85"><output class="ld-slider__output ld-slider__output--permanent" htmlfor="v1" style="--c: var(--v1);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders 2 thumbs 1`] = `
<ld-slider class="ld-slider" role="group" value="10,50" style="--v0: 10; --v1: 50; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v1) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      From
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="10"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <label class="sr-only" htmlfor="v1">
      To
    </label>
    <input class="ld-slider__input" id="v1" max="100" min="0" type="range" value="50"><output class="ld-slider__output" htmlfor="v1" style="--c: var(--v1);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders 2 thumbs with custom thumb labels 1`] = `
<ld-slider class="ld-slider" role="group" value="10,50" style="--v0: 10; --v1: 50; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v1) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Von
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="10"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <label class="sr-only" htmlfor="v1">
      Bis
    </label>
    <input class="ld-slider__input" id="v1" max="100" min="0" type="range" value="50"><output class="ld-slider__output" htmlfor="v1" style="--c: var(--v1);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders aria-disabled 1`] = `
<ld-slider class="ld-slider ld-slider--disabled" role="group" value="0" style="--v0: 0; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input aria-disabled="true" class="ld-slider__input" id="v0" max="100" min="0" type="range" value="0"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders custom edges 1`] = `
<ld-slider class="ld-slider" role="group" value="-50" style="--v0: -50; --ld-slider-width: 100%; --min: -50; --max: 50; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input class="ld-slider__input" id="v0" max="50" min="-50" type="range" value="-50"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: -50;">
      -50
    </div>
    <div class="ld-slider__edge" style="--c: -25;">
      -25
    </div>
    <div class="ld-slider__edge" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge" style="--c: 25;">
      25
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 50;">
      50
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders default 1`] = `
<ld-slider class="ld-slider" role="group" value="0" style="--v0: 0; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="0"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders disabled 1`] = `
<ld-slider class="ld-slider ld-slider--disabled" role="group" value="0" style="--v0: 0; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input aria-disabled="true" class="ld-slider__input" disabled="" id="v0" max="100" min="0" type="range" value="0"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders more than 2 thumbs with custom thumb labels 1`] = `
<ld-slider class="ld-slider" role="group" value="10,50,80" style="--v0: 10; --v1: 50; --v2: 80; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v1) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v2) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Thumb 1
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="10"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <label class="sr-only" htmlfor="v1">
      Thumb 2
    </label>
    <input class="ld-slider__input" id="v1" max="100" min="0" type="range" value="50"><output class="ld-slider__output" htmlfor="v1" style="--c: var(--v1);"></output>
    <label class="sr-only" htmlfor="v2">
      Thumb 3
    </label>
    <input class="ld-slider__input" id="v2" max="100" min="0" type="range" value="80"><output class="ld-slider__output" htmlfor="v2" style="--c: var(--v2);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders with custom radix 1`] = `
<ld-slider class="ld-slider" role="group" value="0" style="--v0: 0; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="0"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders with custom width 1`] = `
<ld-slider class="ld-slider" role="group" value="0" style="--v0: 0; --ld-slider-width: 500px; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="0"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders with inverted color indicators 1`] = `
<ld-slider class="ld-slider" role="group" value="0" style="--v0: 0; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red 100%,
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="0"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider renders with values always visible 1`] = `
<ld-slider class="ld-slider" role="group" value="0" style="--v0: 0; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="0"><output class="ld-slider__output ld-slider__output--permanent" htmlfor="v0" style="--c: var(--v0);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider strict mode corrects invalid initial values 1`] = `
<ld-slider class="ld-slider" role="group" value="40,50,90,100" style="--v0: 40; --v1: 50; --v2: 90; --v3: 100; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v1) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v2) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v3) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="40"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <label class="sr-only" htmlfor="v1">
      Value 2
    </label>
    <input class="ld-slider__input" id="v1" max="100" min="0" type="range" value="50"><output class="ld-slider__output" htmlfor="v1" style="--c: var(--v1);"></output>
    <label class="sr-only" htmlfor="v2">
      Value 3
    </label>
    <input class="ld-slider__input" id="v2" max="100" min="0" type="range" value="90"><output class="ld-slider__output" htmlfor="v2" style="--c: var(--v2);"></output>
    <label class="sr-only" htmlfor="v3">
      Value 4
    </label>
    <input class="ld-slider__input" id="v3" max="100" min="0" type="range" value="100"><output class="ld-slider__output" htmlfor="v3" style="--c: var(--v3);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider strict mode prevents invalid value changes without re-rendering 1`] = `
<ld-slider class="ld-slider" role="group" value="51,50" style="--v0: 0; --v1: 100; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v1) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      From
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="0"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <label class="sr-only" htmlfor="v1">
      To
    </label>
    <input class="ld-slider__input" id="v1" max="100" min="0" type="range" value="100"><output class="ld-slider__output" htmlfor="v1" style="--c: var(--v1);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider strict mode prevents thumbs from being swapped 1`] = `
<ld-slider class="ld-slider" role="group" value="60,60" style="--v0: 60; --v1: 60; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
),
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v1) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      From
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="60"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <label class="sr-only" htmlfor="v1">
      To
    </label>
    <input class="ld-slider__input" id="v1" max="100" min="0" type="range" value="60"><output class="ld-slider__output" htmlfor="v1" style="--c: var(--v1);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;

exports[`ld-slider updates css props on value changes 1`] = `
<ld-slider class="ld-slider" role="group" value="50" style="--v0: 50; --ld-slider-width: 100%; --min: 0; --max: 100; --fill: 
linear-gradient(
  90deg,
  red
    calc(
      var(--ld-slider-radius) + (var(--v0) - var(--min)) /
        var(--ld-slider-diff) * var(--ld-slider-useful-width)
    ),
  transparent 0
);">
  <mock:shadow-root>
    <label class="sr-only" htmlfor="v0">
      Value 1
    </label>
    <input class="ld-slider__input" id="v0" max="100" min="0" type="range" value="50"><output class="ld-slider__output" htmlfor="v0" style="--c: var(--v0);"></output>
    <div class="ld-slider__edge ld-slider__edge--first" style="--c: 0;">
      0
    </div>
    <div class="ld-slider__edge ld-slider__edge--last" style="--c: 100;">
      100
    </div>
  </mock:shadow-root>
</ld-slider>
`;
